tests/functions/test_user_defined_func.py

```diff
@@ -318,25 +355,38 @@ def myFunction(variable):
         val = benchmark(e, 0)
         assert np.allclose(val, expected)
 
-    @pytest.mark.parametrize("op,expected", [ # parameter is string since compiled udf doesn't support closures as of present
-                    ("TANH", [0.76159416, 0.99505475]),
-                    ("EXP", [2.71828183, 20.08553692]),
+    @pytest.mark.parametrize("op,variable,expected", [ # parameter is string since compiled udf doesn't support closures as of present
+                    ("TANH", [[1, 3]], [0.76159416, 0.99505475]),
+                    ("EXP", [[1, 3]], [2.71828183, 20.08553692]),
+                    ("SHAPE", [1, 2], [2]),
+                    ("SHAPE", [[1, 3]], [1, 2]),
+                    ("ASTYPE_FLOAT", [1], [1.0]),
+                    ("ASTYPE_INT", [-1.5], [-1.0]),
                     ])
     @pytest.mark.parametrize("bin_execute", ['Python',
                                              pytest.param('LLVM', marks=pytest.mark.llvm),
                                              pytest.param('PTX', marks=[pytest.mark.llvm, pytest.mark.cuda]),
                                             ])
     @pytest.mark.benchmark(group="Function UDF")
-    def test_user_def_func_numpy(self, op, expected, bin_execute, benchmark):
-        variable = [[1, 3]]
+    def test_user_def_func_numpy(self, op, variable, expected, bin_execute, benchmark):
         if op == "TANH":
             def myFunction(variable):
                 return np.tanh(variable)
         elif op == "EXP":
             def myFunction(variable):
                 return np.exp(variable)
+        elif op == "SHAPE":
+            def myFunction(variable):
+                return variable.shape
+        elif op == "ASTYPE_FLOAT":
+            def myFunction(variable):
+                return variable.astype(float)
```

Comment:
Can input types be integer? or does this suffer from the same issue as the variant below?

Reply:
indeed - this suffers the same issue, where the input type can only be float.

Reply:
so this version only tests the no-op float->float variant and the one below tests float->int->float.
I think int->int no-op should also be tested.

Reply:
not sure how this can be done. is there a way you had in mind to do this?

Reply:
should we just drop the support for int->int casting then? if it's tricky to use, it might not be needed.

Reply:
From a user standpoint, it would be extremely confusing to have to carefully manage types from within Python (e.g. udf returns int, and downstream, multiplies by a float. this would throw an opaque error with no clear fix).

It may be best to restrict udfs to only consider floats - that is, by casting any numbers that are integer constants into floats automatically (potentially warning the user if there are negative side-effects).

Reply:
Why would they need to 'carefully manage types' ? Either int->int casts are easy to have in UDF and should be tested, or it's an extreme case that is difficult to hit and can be dropped.

Reply:
removed int->int casts